src/GraphQL/Execution/ExecutionStrategy.cs

```diff
@@ -357,15 +357,49 @@ void SetArrayItemNode(object d)
                         }
                         else if (node is ValueExecutionNode valueNode)
                         {
-                            node.Result = valueNode.GraphType.Serialize(d)
-                                ?? throw new InvalidOperationException($"Unable to serialize '{d}' to '{valueNode.GraphType.Name}' for list index {index}.");
+                            object serializedResult = null;
+
+                            try
+                            {
+                                serializedResult = valueNode.GraphType.Serialize(d);
+                            }
+                            catch (Exception ex)
+                            {
+                                // Stop execution as early as possible, it doesn't make sense to set the rest of the array elements and then execute nodes
+                                if (listType.ResolvedType is NonNullGraphType)
+                                {
+                                    throw new InvalidOperationException($"Unable to serialize '{d}' to '{valueNode.GraphType.Name}' for list index {index}.", ex);
+                                }
+                                else
+                                {
+                                    SetNodeError(context, node, new ExecutionError($"Unable to serialize '{d}' to '{valueNode.GraphType.Name}'.", ex));
+                                    arrayItems.Add(node);
+                                    return;
+                                }
```

Comment:
Otherwise, the error should be wrapped by `ProcessNodeUnhandledException`, which depending on the execution options may instruct the exception to be thrown.

Reply:
If you agree, let's merge the code suggestion, or wait on resolving the conversation until we agree on a resolution.